opencog/nlp/fuzzy/FuzzySCM.cc

```diff
@@ -0,0 +1,132 @@
+/*
+ * FuzzySCM.cc
+ *
+ * Copyright (C) 2015 OpenCog Foundation
+ *
+ * Author: Leung Man Hin <https://github.com/leungmanhin>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License v3 as
+ * published by the Free Software Foundation and including the exceptions
+ * at http://opencog.org/wiki/Licenses
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program; if not, write to:
+ * Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include <opencog/atoms/pattern/PatternLink.h>
+#include <opencog/guile/SchemePrimitive.h>
+#include <opencog/nlp/types/atom_types.h>
+
+#include "FuzzySCM.h"
+#include "Fuzzy.h"
+
+using namespace opencog::nlp;
+using namespace opencog;
+
+/**
+ * The constructor for FuzzySCM.
+ */
+FuzzySCM::FuzzySCM()
+{
+    static bool is_init = false;
+    if (is_init) return;
+    is_init = true;
+    scm_with_guile(init_in_guile, this);
+}
+
+/**
+ * Init function for using with scm_with_guile.
+ *
+ * Creates the fuzzy scheme module and uses it by default.
+ *
+ * @param self  pointer to the FuzzySCM object
+ */
+void* FuzzySCM::init_in_guile(void* self)
+{
+    scm_c_define_module("opencog nlp fuzzy", init_in_module, self);
+    scm_c_use_module("opencog nlp fuzzy");
+    return NULL;
+}
+
+/**
+ * The main function for defining stuff in the fuzzy scheme module.
+ *
+ * @param data  pointer to the FuzzySCM object
+ */
+void FuzzySCM::init_in_module(void* data)
+{
+    FuzzySCM* self = (FuzzySCM*) data;
+    self->init();
+}
+
+/**
+ * The main init function for the FuzzySCM object.
+ */
+void FuzzySCM::init()
+{
+#ifdef HAVE_GUILE
+    define_scheme_primitive("nlp-fuzzy-match", &FuzzySCM::do_nlp_fuzzy_match,
+                            this, "nlp fuzzy");
+#endif
+}
+
+/**
+ * Implement the "nlp-fuzzy-match" scheme primitive. It calls the nlp fuzzy
+ * matcher to search for potential solutions.
+ *
+ * @param pat        The input pattern
+ * @param rtn_type   The return type (the type of atom we are looking for)
+ * @param excl_list  The exclude-list, containing a list of atoms that we
+ *                   don't want in the result set
+ * @return           A list of solutions and their similarity scores
+ */
+Handle FuzzySCM::do_nlp_fuzzy_match(Handle pat, Type rtn_type,
+                                    const HandleSeq& excl_list,
+                                    bool dup_check)
```

Comment:
ahh .. I see. This is the complicated API that pull req https://github.com/opencog/atomspace/pull/546 is for.  

It would be better if "exclude_list" as actually part of the pattern, by using AbsentLink.

I also really really need a "must-have" list, of terms that must appear in the pattern, exactly as they are, no fuzziness. It would be best if this was a part of the pattern, as well.  Not sure how that could work.   I'm wriing an email now, to suggest an idea...